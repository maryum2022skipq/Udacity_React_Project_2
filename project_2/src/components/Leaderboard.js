import Navbar from "./Navbar";
import { connect } from "react-redux";

const Leaderboard = (props) => {
  return (
    <div>
      <Navbar />
      <div className="leaderboard_table_container">
        <table>
          <thead>
            <th>Users</th>
            <th>Answered</th>
            <th>Created</th>
          </thead>
          <tbody>
            {props.countList.map((user) => {
              return (
                <tr key={user.id}>
                  <td>
                    <img src={user.avatarURL} alt="Avatar" className="avatar" />
                    <br />
                    {user.name} <br />
                    {user.id}
                  </td>
                  <td>{user.answered}</td>
                  <td>{user.created}</td>
                </tr>
              );
            })}
          </tbody>
        </table>
      </div>
    </div>
  );
};
const mapStateToProps = ({ users }) => {
  const userList = Object.keys(users).sort(function (a, b) {
    return (
      Object.keys(users[b].answers).length +
      users[b].questions.length -
      (Object.keys(users[a].answers).length + users[a].questions.length)
    );
  });
  const countList = userList.map((user) => {
    return {
      id: users[user].id,
      name: users[user].name,
      avatarURL: users[user].avatarURL,
      created: Object.keys(users[user].questions).length,
      answered: Object.keys(users[user].answers).length,
    };
  });
  return {
    countList,
  };
};
export default connect(mapStateToProps)(Leaderboard);
