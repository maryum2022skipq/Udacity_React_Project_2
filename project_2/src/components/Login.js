import React from "react";
import { useState } from "react";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";
import { useNavigate } from "react-router-dom";

function Login(props) {
  const [user, setUser] = useState("");
  const navigate = useNavigate();
  const handleChange = (e) => {
    setUser(e.target.value);
  };

  const handleSubmit = () => {
    props.dispatch(setAuthedUser(user));
    if (window.location.pathname === "/") {
      navigate("/");
    }
  };

  return (
    <div>
      <h1>Employee Polls</h1>
      <img
        src="https://png.pngtree.com/png-vector/20191003/ourmid/pngtree-user-login-or-authenticate-icon-on-gray-background-flat-icon-ve-png-image_1786166.jpg"
        alt="loginPage"
      />
      <h2>Login</h2>
      <div>
        <h4>User</h4>
        <select
          onChange={handleChange}
          data-testid="login_select"
          defaultValue={"none"}
        >
          <option value="none" disabled>
            Select User
          </option>
          {props.users
            ? props.users.map((user) => {
                return (
                  <option key={user} value={user}>
                    {user}
                  </option>
                );
              })
            : ""}
        </select>
      </div>
      <input
        type="submit"
        value="LOGIN"
        data-testid="login_submit"
        className="btn"
        onClick={handleSubmit}
        disabled={user === ""}
      />
    </div>
  );
}

const mapStateToProps = ({ users }) => {
  return {
    users: Object.keys(users),
  };
};

export default connect(mapStateToProps)(Login);
