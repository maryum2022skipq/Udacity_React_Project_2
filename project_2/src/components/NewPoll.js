import Navbar from "./Navbar";
import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { handleSaveQuestion } from "../actions/questions";
import { useNavigate } from "react-router-dom";

const NewPoll = (props) => {
  const [optionOne, setOptionOne] = useState("");
  const [optionTwo, setOptionTwo] = useState("");
  const [isDisabled, setDisabled] = useState(true);

  const navigate = useNavigate();
  const handleSubmit = (e) => {
    e.preventDefault();
    props.dispatch(handleSaveQuestion(props.authedUser, optionOne, optionTwo));
    navigate("/");
  };
  useEffect(() => {
    const check_inputFields = () => {
      if (optionOne.length > 0 && optionTwo.length > 0) {
        setDisabled(false);
      }
    };
    check_inputFields();
  }, [optionOne, optionTwo]);

  return (
    <div>
      <Navbar />
      <div>
        <h2>Would You Rather</h2>
        <h5 style={{ color: "grey" }}>Create Your Own Poll</h5>
      </div>
      <form>
        <div>
          <label>First Option</label>
          <div>
            <input
              className="new_poll_input"
              type="text"
              data-testid="optionOne"
              placeholder="Option 1"
              onChange={(e) => {
                setOptionOne(e.target.value);
              }}
            />
          </div>
        </div>
        <div>
          <label>Second Option</label>
          <div>
            <input
              className="new_poll_input"
              type="text"
              data-testid="optionTwo"
              placeholder="Option 2"
              onChange={(e) => {
                setOptionTwo(e.target.value);
              }}
            />
          </div>
        </div>
        <input
          type="submit"
          value="Submit"
          data-testid="submit_poll"
          disabled={isDisabled}
          className="btn"
          onClick={handleSubmit}
        />
      </form>
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  authedUser,
});
export default connect(mapStateToProps)(NewPoll);
