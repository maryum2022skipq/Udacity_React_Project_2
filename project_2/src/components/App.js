import { useEffect } from "react";
import { connect } from "react-redux";
import { handleInitialData } from "../actions/shared";
import { Routes, Route } from "react-router-dom";
import Login from "./Login";
import LoadingBar from "react-redux-loading-bar";
import Home from "./Home";
import PageNotFound from "./PageNotFound";
import NewPoll from "./NewPoll";
import Leaderboard from "./Leaderboard";
import PollDetail from "./PollDetail";

const App = (props) => {
  useEffect(() => {
    props.dispatch(handleInitialData());
  }, [props]);

  return (
    <div className="App">
      <LoadingBar />
      {props.authedUser ? (
        <Routes>
          <Route exact path="/" element={<Home />} />
          <Route path="*" element={<PageNotFound />} />
          <Route path="/add" element={<NewPoll />} />
          <Route path="/leaderboard" element={<Leaderboard />} />
          <Route path="/questions/:question_id" element={<PollDetail />} />
        </Routes>
      ) : (
        <Login />
      )}
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  loading: authedUser === null,
  authedUser,
});

export default connect(mapStateToProps)(App);
