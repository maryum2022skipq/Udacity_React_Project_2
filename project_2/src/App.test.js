import { render, fireEvent } from "@testing-library/react";
import App from "../src/components/App";
import { _saveQuestion, _saveQuestionAnswer } from "./utils/_DATA";
import { BrowserRouter } from "react-router-dom";
import { createStore } from "redux";
import { Provider } from "react-redux";
import reducer from "./reducers";
import thunk from "redux-thunk";
import { applyMiddleware } from "redux";
import Leaderboard from "./components/Leaderboard";
import NewPoll from "./components/NewPoll";
import Login from "./components/Login";
//test("renders App", () => {
//  render(<App />);
//  screen.debug();
//});

describe("async unit test for _saveQuestion", () => {
  it("verify that the saved question is returned with all expected fields ", async () => {
    const optionOneText = "Eat Pizza";
    const optionTwoText = "Eat Burger";
    const author = "mtsamis";
    const question = {
      optionOneText,
      optionTwoText,
      author,
    };
    const answer = await _saveQuestion(question);
    expect(answer.author).toEqual("mtsamis");
    expect(answer.optionOne.text).toEqual(optionOneText);
    expect(answer.optionTwo.text).toEqual(optionTwoText);
  });
});

describe("async unit test for _saveQuestion", () => {
  it("verify that an error is returned if incorrect data is passed", async () => {
    const optionOneText = "Eat Pizza";
    const optionTwoText = "Eat Burger";
    const question = {
      optionOneText,
      optionTwoText,
    };
    await expect(_saveQuestion(question)).rejects.toEqual(
      "Please provide optionOneText, optionTwoText, and author"
    );
  });
});

describe("async unit test for _saveQuestionAnswer", () => {
  it("verify that the saved question answer is returned with all expected fields ", async () => {
    const authedUser = "sarahedo";
    const qid = "vthrdm985a262al8qx3do";
    const answer = "optionTwo";
    return await expect(
      _saveQuestionAnswer({
        authedUser,
        qid,
        answer,
      })
    ).resolves.toBe(true);
  });
});

describe("async unit test for _saveQuestionAnswer", () => {
  it("verify that an error is returned if incorrect data is passed", async () => {
    const authedUser = "sarahedo";
    const answer = "optionOne";
    await expect(
      _saveQuestionAnswer({
        authedUser,
        answer,
      })
    ).rejects.toEqual("Please provide authedUser, qid, and answer");
  });
});

const store = createStore(reducer, applyMiddleware(thunk));

describe("a snapshot test for app component", () => {
  it("verify that app component renders properly and matches with the snapshot", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});

describe("a snapshot test for login component", () => {
  test("verify that login component renders properly and matches with the snapshot ", () => {
    const component = render(
      <Provider store={store} dispatch={store.dispatch}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});

describe("a DOM query test ", () => {
  test("verify that login component renders dropdown menu to select user", () => {
    const component = render(
      <Provider store={store} dispatch={store.dispatch}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    var user = component.getByTestId("login_select");
    expect(user).toBeInTheDocument();
  });
});

describe("a DOM query test", () => {
  test("verifies that when login submit button is pressed, home is rendered.", () => {
    const component = render(
      <Provider store={store} dispatch={store.dispatch}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    fireEvent.click(component.getByTestId("login_submit"));
    expect(global.window.location.pathname).toEqual("/");
  });
});
