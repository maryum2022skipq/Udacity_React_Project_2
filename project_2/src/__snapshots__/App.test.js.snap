// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a snapshot test for app component verify that app component renders properly and matches with the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <div
          style="direction: ltr;"
        >
          <div
            class=""
            style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
          />
          <div
            style="display: table; clear: both;"
          />
        </div>
        <div>
          <h1>
            Employee Polls
          </h1>
          <img
            alt="loginPage"
            src="https://png.pngtree.com/png-vector/20191003/ourmid/pngtree-user-login-or-authenticate-icon-on-gray-background-flat-icon-ve-png-image_1786166.jpg"
          />
          <h2>
            Login
          </h2>
          <div>
            <h4>
              User
            </h4>
            <select
              data-testid="login_select"
            >
              <option
                disabled=""
                selected=""
                value="none"
              >
                Select User
              </option>
            </select>
          </div>
          <input
            class="btn"
            data-testid="login_submit"
            disabled=""
            type="submit"
            value="LOGIN"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <div
        style="direction: ltr;"
      >
        <div
          class=""
          style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
        />
        <div
          style="display: table; clear: both;"
        />
      </div>
      <div>
        <h1>
          Employee Polls
        </h1>
        <img
          alt="loginPage"
          src="https://png.pngtree.com/png-vector/20191003/ourmid/pngtree-user-login-or-authenticate-icon-on-gray-background-flat-icon-ve-png-image_1786166.jpg"
        />
        <h2>
          Login
        </h2>
        <div>
          <h4>
            User
          </h4>
          <select
            data-testid="login_select"
          >
            <option
              disabled=""
              selected=""
              value="none"
            >
              Select User
            </option>
          </select>
        </div>
        <input
          class="btn"
          data-testid="login_submit"
          disabled=""
          type="submit"
          value="LOGIN"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`a snapshot test for login component verify that login component renders properly and matches with the snapshot  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1>
          Employee Polls
        </h1>
        <img
          alt="loginPage"
          src="https://png.pngtree.com/png-vector/20191003/ourmid/pngtree-user-login-or-authenticate-icon-on-gray-background-flat-icon-ve-png-image_1786166.jpg"
        />
        <h2>
          Login
        </h2>
        <div>
          <h4>
            User
          </h4>
          <select
            data-testid="login_select"
          >
            <option
              disabled=""
              selected=""
              value="none"
            >
              Select User
            </option>
          </select>
        </div>
        <input
          class="btn"
          data-testid="login_submit"
          disabled=""
          type="submit"
          value="LOGIN"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        Employee Polls
      </h1>
      <img
        alt="loginPage"
        src="https://png.pngtree.com/png-vector/20191003/ourmid/pngtree-user-login-or-authenticate-icon-on-gray-background-flat-icon-ve-png-image_1786166.jpg"
      />
      <h2>
        Login
      </h2>
      <div>
        <h4>
          User
        </h4>
        <select
          data-testid="login_select"
        >
          <option
            disabled=""
            selected=""
            value="none"
          >
            Select User
          </option>
        </select>
      </div>
      <input
        class="btn"
        data-testid="login_submit"
        disabled=""
        type="submit"
        value="LOGIN"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
